File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        unique = set()
        l = r = long = 0
        
        while r < len(s):
            if s[r] in unique:
                long = max(long, r-l)
                while (s[l] != s[r]):
                    unique.remove(s[l])
                    l += 1
                l += 1
            unique.add(s[r])
            r += 1
            
        return max(long, r - l)

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = 0, start = 0, n = s.size();
        unordered_set<char> st;
        
        for (int end = 0; end < n; end++) {
            while (st.count(s[end])) {
                st.erase(s[start]);
                start++;
            }
            st.insert(s[end]);
            ans = max(ans, (int)st.size());
        }
        return ans;
    }
};